pkg/ingester/shipper.go

```diff
@@ -233,41 +237,68 @@ func (s *Shipper) blockMetasFromOldest() (metas []*metadata.Meta, _ error) {
 	return metas, nil
 }
 
-func readShippedBlocks(dir string) (map[ulid.ULID]struct{}, error) {
+func readShippedBlocks(dir string) (map[ulid.ULID]time.Time, error) {
 	meta, err := readShipperMetaFile(dir)
+	if err != nil {
+		return readThanosShippedBlocks(dir)
+	}
+
+	shippedBlocks := make(map[ulid.ULID]time.Time, len(meta.Shipped))
+	for _, shippedBlock := range meta.Shipped {
+		shippedBlocks[shippedBlock.BlockID] = shippedBlock.ShippedTime
+	}
+	return shippedBlocks, nil
+}
+
+// readThanosShippedBlocks reads the legacy thanos.shipper.json file.
+// TODO Remove in Mimir 2.7.0
+func readThanosShippedBlocks(dir string) (map[ulid.ULID]time.Time, error) {
+	meta, err := readThanosShipperMetaFile(dir)
 	if errors.Is(err, os.ErrNotExist) {
 		// If the meta file doesn't exist it means the shipper hasn't run yet.
-		meta = &shipperMeta{}
+		meta = &thanosShipperMeta{}
 	} else if err != nil {
 		return nil, err
 	}
 
-	// Build a map.
-	shippedBlocks := make(map[ulid.ULID]struct{}, len(meta.Uploaded))
+	shippedBlocks := make(map[ulid.ULID]time.Time, len(meta.Uploaded))
 	for _, blockID := range meta.Uploaded {
-		shippedBlocks[blockID] = struct{}{}
+		shippedBlocks[blockID] = ulid.Time(blockID.Time()).Add(estimatedShippingDuration)
 	}
 
 	return shippedBlocks, nil
 }
 
-// shipperMeta defines the format thanos.shipper.json file that the shipper places in the data directory.
-type shipperMeta struct {
+// thanosShipperMeta defines the format thanos.shipper.json file that the shipper places in the data directory.
+type thanosShipperMeta struct {
 	Version  int         `json:"version"`
 	Uploaded []ulid.ULID `json:"uploaded"`
 }
 
+type shippedBlock struct {
+	BlockID     ulid.ULID `json:"block_id"`
+	ShippedTime time.Time `json:"shipped_time"`
+}
+
+// shipperMeta defines the format thanos.shipper.json file that the shipper places in the data directory.
+type shipperMeta struct {
+	Version int            `json:"version"`
+	Shipped []shippedBlock `json:"shipped"`
```

Comment:
What about `Shipped map[ulid.ULID]time.Time` instead? (or perhaps `model.Time`)

Reply:
Using `Shipped map[ulid.ULID]model.Time` looks like this:

```
{
  "version": 1,
  "shipped": {
    "01GPGRWYPY0000000000000000": "2023-01-11T16:59:07.486415+01:00"
  }
}
```

and with `model.Time`:
```
{
  "version": 1,
  "shipped": {
    "01GPGRZ5H70000000000000000": 1673452820.007
  }
}
```

I'd prefer the last version.

Reply:
Okey, used last version.